src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -70,53 +51,25 @@ private NamedTypeSymbol ValueTaskT
     {
         var nodeType = node.Expression.Type;
         Debug.Assert(nodeType is not null);
-        var originalType = nodeType.OriginalDefinition;
-
-        WellKnownMember awaitCall;
-        TypeWithAnnotations? maybeNestedType = null;
 
-        if (ReferenceEquals(originalType, Task))
-        {
-            awaitCall = WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__AwaitTask;
-        }
-        else if (ReferenceEquals(originalType, TaskT))
-        {
-            awaitCall = WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__AwaitTaskT_T;
-            maybeNestedType = ((NamedTypeSymbol)nodeType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
-        }
-        else if (ReferenceEquals(originalType, ValueTask))
-        {
-            awaitCall = WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__AwaitValueTask;
-        }
-        else if (ReferenceEquals(originalType, ValueTaskT))
+        var awaitableInfo = node.AwaitableInfo;
+        var runtimeAsyncAwaitMethod = awaitableInfo.RuntimeAsyncAwaitMethod;
+        Debug.Assert(runtimeAsyncAwaitMethod is not null);
+        Debug.Assert(ReferenceEquals(
+            runtimeAsyncAwaitMethod.ContainingType.OriginalDefinition,
+            _factory.Compilation.GetSpecialType(InternalSpecialType.System_Runtime_CompilerServices_AsyncHelpers)));
+        Debug.Assert(runtimeAsyncAwaitMethod.Name is "Await" or "UnsafeAwaitAwaiter" or "AwaitAwaiter");
+
+        if (runtimeAsyncAwaitMethod.Name == "Await")
```

Comment:
nit: this feels a bit weird. Would it be better to split this into two fields in `BoundAwaitableInfo` (one for `Await` case and one for `AwaitAwait` case)?

Reply:
I thought about it, but decided I didn't want to have two fields when they were mutually exclusive. I also considered a subclass approach, and felt that was too complicated for the effort. Let me know if you feel strongly about this; I can be persuaded.